crates/uv-resolver/src/lock.rs

```diff
@@ -1,27 +1,133 @@
+    GitSourceDist, IndexUrl, Name, PathBuiltDist, PathSourceDist, RegistryBuiltDist,
+    RegistrySourceDist, Resolution, ResolvedDist, ToUrlError, VersionOrUrl,
 };
 use pep440_rs::Version;
+use pep508_rs::{MarkerEnvironment, VerbatimUrl};
+use platform_tags::{TagCompatibility, TagPriority, Tags};
 use pypi_types::HashDigest;
+use rustc_hash::FxHashMap;
 use url::Url;
+use uv_normalize::PackageName;
 
 #[derive(Clone, Debug)]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[cfg_attr(feature = "serde", serde(into = "LockWire", try_from = "LockWire"))]
 pub struct Lock {
     version: u32,
-    #[cfg_attr(feature = "serde", serde(rename = "distribution"))]
     distributions: Vec<Distribution>,
+    /// A map from distribution ID to index in `distributions`.
+    ///
+    /// This can be used to quickly lookup the full distribution for any ID
+    /// in this lock. For example, the dependencies for each distribution are
+    /// listed as distributions IDs. This map can be used to find the full
+    /// distribution for each such dependency.
+    ///
+    /// It is guaranteed that every distribution in this lock has an entry in
+    /// this map, and that every dependency for every distribution has an ID
+    /// that exists in this map. That is, there are no dependencies that don't
+    /// have a corresponding locked distribution entry in the same lock file.
+    by_id: FxHashMap<DistributionId, usize>,
 }
 
 impl Lock {
-    pub(crate) fn new(mut distributions: Vec<Distribution>) -> Result<Lock, LockError> {
-        for dist in &mut distributions {
+    pub(crate) fn new(distributions: Vec<Distribution>) -> Result<Lock, LockError> {
+        let wire = LockWire {
+            version: 1,
+            distributions,
+        };
+        Lock::try_from(wire)
+    }
+
+    pub fn to_resolution(
+        &self,
+        marker_env: &MarkerEnvironment,
+        tags: &Tags,
+        root_name: &str,
+    ) -> Resolution {
+        let root = self
+            .find_by_name(root_name)
+            // TODO: In the future, we should derive the root distribution
+            // from the pyproject.toml, but I don't think the infrastructure
+            // for that is in place yet. For now, we ask the caller to specify
+            // the root package name explicitly, and we assume here that it is
+            // correct.
+            .expect("found too many distributions matching root")
+            .expect("could not find root");
+        let mut queue: VecDeque<&Distribution> = VecDeque::new();
+        queue.push_back(root);
+
+        let mut map = FxHashMap::default();
+        while let Some(dist) = queue.pop_front() {
+            for dep in &dist.dependencies {
+                let dep_dist = self.find_by_id(&dep.id);
+                queue.push_back(dep_dist);
+            }
+            let name = PackageName::new(dist.id.name.to_string()).unwrap();
```

Comment:
Should we make `DistributionId.name` a `PackageName`?

Reply:
Ah yeah, nice idea. Done.